The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13319433075/job/37201033329) [(plain)](https://github.com/rust-lang-ci/rust/commit/7c77e1ac034e34f6a4e0ea2752b3734e18bcc2bb/checks/37201033329/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen/intrinsics/cold_path3.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/ci-llvm/bin/FileCheck" "--input-file" "/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll" "/Users/runner/work/rust/rust/tests/codegen/intrinsics/cold_path3.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
--- stderr -------------------------------
/Users/runner/work/rust/rust/tests/codegen/intrinsics/cold_path3.rs:77:11: error: CHECK: expected string not found in input
/Users/runner/work/rust/rust/tests/codegen/intrinsics/cold_path3.rs:77:11: error: CHECK: expected string not found in input
// CHECK: ], !prof !3
          ^
/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll:106:19: note: scanning from here
 i32 3, label %bb3
/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll:107:2: note: possible intended match here
/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll:107:2: note: possible intended match here
 ], !prof !2
/Users/runner/work/rust/rust/tests/codegen/intrinsics/cold_path3.rs:84:11: error: CHECK: expected string not found in input
/Users/runner/work/rust/rust/tests/codegen/intrinsics/cold_path3.rs:84:11: error: CHECK: expected string not found in input
// CHECK: ], !prof !5
          ^
/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll:145:20: note: scanning from here
 i32 13, label %bb3
/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll:146:2: note: possible intended match here
/Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll:146:2: note: possible intended match here
 ], !prof !4

Input file: /Users/runner/work/rust/rust/build/aarch64-apple-darwin/test/codegen/intrinsics/cold_path3/cold_path3.ll
Check file: /Users/runner/work/rust/rust/tests/codegen/intrinsics/cold_path3.rs


-dump-input=help explains the following input dump.
Input was:
<<<<<<
            .
            .
            .
            .
            6: @alloc_9a6b3164498723817f82a587f1e4683c = private unnamed_addr constant <{ [7 x i8] }> <{ [7 x i8] c"path a\0A" }>, align 1 
            7: @alloc_1c578f6a7b79fce31d5ef7a209e49338 = private unnamed_addr constant <{ ptr, [8 x i8] }> <{ ptr @alloc_9a6b3164498723817f82a587f1e4683c, [8 x i8] c"\07\00\00\00\00\00\00\00" }>, align 8 
            8: @alloc_e5259ec2649ced3f287f8fd7e7344232 = private unnamed_addr constant <{ [7 x i8] }> <{ [7 x i8] c"path b\0A" }>, align 1 
            9: @alloc_3dcc9f1605c7ab420320f7000b1f4afd = private unnamed_addr constant <{ ptr, [8 x i8] }> <{ ptr @alloc_e5259ec2649ced3f287f8fd7e7344232, [8 x i8] c"\07\00\00\00\00\00\00\00" }>, align 8 
           10: @alloc_62aaa2879f291e69e6cf027569d76347 = private unnamed_addr constant <{ [7 x i8] }> <{ [7 x i8] c"path c\0A" }>, align 1 
           11: @alloc_67d34dcbe327fb4cf5fc5b15cdeaa14e = private unnamed_addr constant <{ ptr, [8 x i8] }> <{ ptr @alloc_62aaa2879f291e69e6cf027569d76347, [8 x i8] c"\07\00\00\00\00\00\00\00" }>, align 8 
           12: @alloc_c4ff755b2c0c8b58fe7e5f24a51d3893 = private unnamed_addr constant <{ [7 x i8] }> <{ [7 x i8] c"path d\0A" }>, align 1 
           13: @alloc_5e8e5ea12fa49d26166c5f3ad659b264 = private unnamed_addr constant <{ ptr, [8 x i8] }> <{ ptr @alloc_c4ff755b2c0c8b58fe7e5f24a51d3893, [8 x i8] c"\07\00\00\00\00\00\00\00" }>, align 8 
           14:  
           15: ; Function Attrs: noinline uwtable 
           16: define void @path_a() unnamed_addr #0 { 
           18:  %_2 = alloca [48 x i8], align 8 
           18:  %_2 = alloca [48 x i8], align 8 
           19:  call void @llvm.lifetime.start.p0(i64 48, ptr nonnull %_2) 
           20:  store ptr @alloc_1c578f6a7b79fce31d5ef7a209e49338, ptr %_2, align 8 
           21:  %0 = getelementptr inbounds i8, ptr %_2, i64 8 
           22:  store i64 1, ptr %0, align 8 
           23:  %1 = getelementptr inbounds i8, ptr %_2, i64 32 
           24:  store ptr null, ptr %1, align 8 
           25:  %2 = getelementptr inbounds i8, ptr %_2, i64 16 
           26:  store ptr inttoptr (i64 8 to ptr), ptr %2, align 8 
           27:  %3 = getelementptr inbounds i8, ptr %_2, i64 24 
           28:  store i64 0, ptr %3, align 8 
           29: ; call std::io::stdio::_print 
           30:  call void @_ZN3std2io5stdio6_print17hd36f514a32cc0647E(ptr noalias nocapture noundef nonnull align 8 dereferenceable(48) %_2) 
           31:  call void @llvm.lifetime.end.p0(i64 48, ptr nonnull %_2) 
           32:  ret void 
           33: } 
           34:  
           35: ; Function Attrs: noinline uwtable 
           36: define void @path_b() unnamed_addr #0 { 
           38:  %_2 = alloca [48 x i8], align 8 
           38:  %_2 = alloca [48 x i8], align 8 
           39:  call void @llvm.lifetime.start.p0(i64 48, ptr nonnull %_2) 
           40:  store ptr @alloc_3dcc9f1605c7ab420320f7000b1f4afd, ptr %_2, align 8 
           41:  %0 = getelementptr inbounds i8, ptr %_2, i64 8 
           42:  store i64 1, ptr %0, align 8 
           43:  %1 = getelementptr inbounds i8, ptr %_2, i64 32 
           44:  store ptr null, ptr %1, align 8 
           45:  %2 = getelementptr inbounds i8, ptr %_2, i64 16 
           46:  store ptr inttoptr (i64 8 to ptr), ptr %2, align 8 
           47:  %3 = getelementptr inbounds i8, ptr %_2, i64 24 
           48:  store i64 0, ptr %3, align 8 
           49: ; call std::io::stdio::_print 
           50:  call void @_ZN3std2io5stdio6_print17hd36f514a32cc0647E(ptr noalias nocapture noundef nonnull align 8 dereferenceable(48) %_2) 
           51:  call void @llvm.lifetime.end.p0(i64 48, ptr nonnull %_2) 
           52:  ret void 
           53: } 
           54:  
           55: ; Function Attrs: noinline uwtable 
           56: define void @path_c() unnamed_addr #0 { 
           58:  %_2 = alloca [48 x i8], align 8 
           58:  %_2 = alloca [48 x i8], align 8 
           59:  call void @llvm.lifetime.start.p0(i64 48, ptr nonnull %_2) 
           60:  store ptr @alloc_67d34dcbe327fb4cf5fc5b15cdeaa14e, ptr %_2, align 8 
           61:  %0 = getelementptr inbounds i8, ptr %_2, i64 8 
           62:  store i64 1, ptr %0, align 8 
           63:  %1 = getelementptr inbounds i8, ptr %_2, i64 32 
           64:  store ptr null, ptr %1, align 8 
           65:  %2 = getelementptr inbounds i8, ptr %_2, i64 16 
           66:  store ptr inttoptr (i64 8 to ptr), ptr %2, align 8 
           67:  %3 = getelementptr inbounds i8, ptr %_2, i64 24 
           68:  store i64 0, ptr %3, align 8 
           69: ; call std::io::stdio::_print 
           70:  call void @_ZN3std2io5stdio6_print17hd36f514a32cc0647E(ptr noalias nocapture noundef nonnull align 8 dereferenceable(48) %_2) 
           71:  call void @llvm.lifetime.end.p0(i64 48, ptr nonnull %_2) 
           72:  ret void 
           73: } 
           74:  
           75: ; Function Attrs: noinline uwtable 
           76: define void @path_d() unnamed_addr #0 { 
           78:  %_2 = alloca [48 x i8], align 8 
           78:  %_2 = alloca [48 x i8], align 8 
           79:  call void @llvm.lifetime.start.p0(i64 48, ptr nonnull %_2) 
           80:  store ptr @alloc_5e8e5ea12fa49d26166c5f3ad659b264, ptr %_2, align 8 
           81:  %0 = getelementptr inbounds i8, ptr %_2, i64 8 
           82:  store i64 1, ptr %0, align 8 
           83:  %1 = getelementptr inbounds i8, ptr %_2, i64 32 
           84:  store ptr null, ptr %1, align 8 
           85:  %2 = getelementptr inbounds i8, ptr %_2, i64 16 
           86:  store ptr inttoptr (i64 8 to ptr), ptr %2, align 8 
           87:  %3 = getelementptr inbounds i8, ptr %_2, i64 24 
           88:  store i64 0, ptr %3, align 8 
           89: ; call std::io::stdio::_print 
           90:  call void @_ZN3std2io5stdio6_print17hd36f514a32cc0647E(ptr noalias nocapture noundef nonnull align 8 dereferenceable(48) %_2) 
           91:  call void @llvm.lifetime.end.p0(i64 48, ptr nonnull %_2) 
           92:  ret void 
           93: } 
           95: ; Function Attrs: uwtable 
           95: ; Function Attrs: uwtable 
           96: define void @test(i32 noundef range(i32 0, 2) %0, i32 %1) unnamed_addr #1 { 
           97: start: 
           98:  %trunc = trunc nuw i32 %0 to i1 
           99:  br i1 %trunc, label %bb2, label %bb1 
          100:  
          101: bb2: ; preds = %start 
          102:  switch i32 %1, label %bb1 [ 
          103:  i32 0, label %bb6 
          104:  i32 1, label %bb5 
          105:  i32 2, label %bb4 
          106:  i32 3, label %bb3 
check:77'0                       X error: no match found
          107:  ], !prof !2 
check:77'0     ~~~~~~~~~~~~~
check:77'1      ?            possible intended match
check:77'0     ~
check:77'0     ~
          109: bb1: ; preds = %start, %bb2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          110:  tail call void @path_a() 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          111:  br label %bb9 
check:77'0     ~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          113: bb6: ; preds = %bb2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~
          114:  tail call void @path_a() 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          115:  br label %bb9 
check:77'0     ~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          117: bb5: ; preds = %bb2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~
          118:  tail call void @path_b() 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          119:  br label %bb9 
check:77'0     ~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          121: bb4: ; preds = %bb2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~
          122:  tail call void @path_c() 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          123:  br label %bb9 
check:77'0     ~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          125: bb3: ; preds = %bb2 
check:77'0     ~~~~~~~~~~~~~~~~~~~~
          126:  tail call void @path_d() 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          127:  br label %bb9 
check:77'0     ~~~~~~~~~~~~~~~
check:77'0     ~
check:77'0     ~
          129: bb9: ; preds = %bb1, %bb3, %bb4, %bb5, %bb6 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          130:  ret void 
check:77'0     ~~~~~~~~~~
          131: } 
check:77'0     ~~
check:77'0     ~
          133: ; Function Attrs: uwtable 
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
check:77'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          134: define void @test2(i32 noundef range(i32 0, 2) %0, i32 %1) unnamed_addr #1 { 
check:77'0     ~~~~~~~~~~~~~~~~~~~
          135: start: 
          136:  %trunc = trunc nuw i32 %0 to i1 
          137:  br i1 %trunc, label %bb2, label %bb1, !prof !3 
          138:  
          139: bb2: ; preds = %start 
          140:  %_6 = icmp ne i32 %1, 12 
          141:  tail call void @llvm.assume(i1 %_6) 
          142:  switch i32 %1, label %bb1 [ 
          143:  i32 10, label %bb5 
          144:  i32 11, label %bb4 
          145:  i32 13, label %bb3 
check:84'0                        X error: no match found
          146:  ], !prof !4 
check:84'0     ~~~~~~~~~~~~~
check:84'1      ?            possible intended match
check:84'0     ~
check:84'0     ~
          148: bb1: ; preds = %start, %bb2 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          149:  tail call void @path_a() 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          150:  br label %bb9 
check:84'0     ~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          152: bb5: ; preds = %bb2 
check:84'0     ~~~~~~~~~~~~~~~~~~~~
          153:  tail call void @path_a() 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          154:  br label %bb9 
check:84'0     ~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          156: bb4: ; preds = %bb2 
check:84'0     ~~~~~~~~~~~~~~~~~~~~
          157:  tail call void @path_b() 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          158:  br label %bb9 
check:84'0     ~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          160: bb3: ; preds = %bb2 
check:84'0     ~~~~~~~~~~~~~~~~~~~~
          161:  tail call void @path_d() 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          162:  br label %bb9 
check:84'0     ~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          164: bb9: ; preds = %bb1, %bb3, %bb4, %bb5 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          165:  ret void 
check:84'0     ~~~~~~~~~~
          166: } 
check:84'0     ~~
check:84'0     ~
          168: ; std::io::stdio::_print 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
          169: ; Function Attrs: uwtable 
          169: ; Function Attrs: uwtable 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~
          170: declare void @_ZN3std2io5stdio6_print17hd36f514a32cc0647E(ptr noalias nocapture noundef align 8 dereferenceable(48)) unnamed_addr #1 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          172: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          173: declare void @llvm.assume(i1 noundef) #2 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          175: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          176: declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #3 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          178: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          179: declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #3 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          181: attributes #0 = { noinline uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="apple-m1" } 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          182: attributes #1 = { uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="apple-m1" } 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          183: attributes #2 = { mustprogress nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) } 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          184: attributes #3 = { mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) } 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          186: !llvm.module.flags = !{!0} 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~
          187: !llvm.ident = !{!1} 
check:84'0     ~~~~~~~~~~~~~~~~~~~~
check:84'0     ~
check:84'0     ~
          189: !0 = !{i32 8, !"PIC Level", i32 2} 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          190: !1 = !{!"rustc version 1.86.0-nightly (7c77e1ac0 2025-02-13)"} 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          191: !2 = !{!"branch_weights", i32 2000, i32 2000, i32 1, i32 2000, i32 1} 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          192: !3 = !{!"branch_weights", i32 2000, i32 2} 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          193: !4 = !{!"branch_weights", i32 1, i32 2000, i32 1, i32 1} 
check:84'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
------------------------------------------



```

</details>

--- a/compiler/rustc_codegen_llvm/src/builder.rs
+++ b/compiler/rustc_codegen_llvm/src/builder.rs
@@ -4,7 +4,7 @@ use std::{iter, ptr};
 
 pub(crate) mod autodiff;
 
-use libc::{c_char, c_uint};
+use libc::{c_char, c_uint, size_t};
 use rustc_abi as abi;
 use rustc_abi::{Align, Size, WrappingRange};
 use rustc_codegen_ssa::MemFlags;
@@ -32,7 +32,7 @@ use crate::abi::FnAbiLlvmExt;
 use crate::attributes;
 use crate::common::Funclet;
 use crate::context::{CodegenCx, SimpleCx};
-use crate::llvm::{self, AtomicOrdering, AtomicRmwBinOp, BasicBlock, False, True};
+use crate::llvm::{self, AtomicOrdering, AtomicRmwBinOp, BasicBlock, False, Metadata, True};
 use crate::type_::Type;
 use crate::type_of::LayoutLlvmExt;
 use crate::value::Value;
@@ -330,6 +330,50 @@ impl<'a, 'll, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
         for (on_val, dest) in cases {
             let on_val = self.const_uint_big(self.val_ty(v), on_val);
             unsafe { llvm::LLVMAddCase(switch, on_val, dest) }
+        }
+    }
+
+    fn switch_with_weights(
+        &mut self,
+        v: Self::Value,
+        else_llbb: Self::BasicBlock,
+        else_is_cold: bool,
+        cases: impl ExactSizeIterator<Item = (u128, Self::BasicBlock, bool)>,
+    ) {
+        if self.cx.sess().opts.optimize == rustc_session::config::OptLevel::No {
+            self.switch(v, else_llbb, cases.map(|(val, dest, _)| (val, dest)));
+            return;
+        }
+
+        let id_str = "branch_weights";
+        let id = unsafe {
+            llvm::LLVMMDStringInContext2(self.cx.llcx, id_str.as_ptr().cast(), id_str.len())
+        };
+
+        // For switch instructions with 2 targets, the `llvm.expect` intrinsic is used.
+        // This function handles switch instructions with more than 2 targets and it needs to
+        // emit branch weights metadata instead of using the intrinsic.
+        // The values 1 and 2000 are the same as the values used by the `llvm.expect` intrinsic.
+        let cold_weight = unsafe { llvm::LLVMValueAsMetadata(self.cx.const_u32(1)) };
+        let hot_weight = unsafe { llvm::LLVMValueAsMetadata(self.cx.const_u32(2000)) };
+        let weight =
+            |is_cold: bool| -> &Metadata { if is_cold { cold_weight } else { hot_weight } };
+
+        let mut md: SmallVec<[&Metadata; 16]> = SmallVec::with_capacity(cases.len() + 2);
+        md.push(id);
+        md.push(weight(else_is_cold));
+
+        let switch =
+            unsafe { llvm::LLVMBuildSwitch(self.llbuilder, v, else_llbb, cases.len() as c_uint) };
+        for (on_val, dest, is_cold) in cases {
+            let on_val = self.const_uint_big(self.val_ty(v), on_val);
+            unsafe { llvm::LLVMAddCase(switch, on_val, dest) }
+            md.push(weight(is_cold));
+        }
+
+        unsafe {
+            let md_node = llvm::LLVMMDNodeInContext2(self.cx.llcx, md.as_ptr(), md.len() as size_t);
+            self.cx.set_metadata(switch, llvm::MD_prof, md_node);
         }
     }
 

--- a/compiler/rustc_codegen_ssa/src/mir/block.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/block.rs
@@ -429,11 +429,34 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
             let cmp = bx.icmp(IntPredicate::IntEQ, discr_value, llval);
             bx.cond_br(cmp, ll1, ll2);
         } else {
-            bx.switch(
-                discr_value,
-                helper.llbb_with_cleanup(self, targets.otherwise()),
-                target_iter.map(|(value, target)| (value, helper.llbb_with_cleanup(self, target))),
-            );
+            let otherwise = targets.otherwise();
+            let otherwise_cold = self.cold_blocks[otherwise];
+            let otherwise_unreachable = self.mir[otherwise].is_empty_unreachable();
+            let cold_count = targets.iter().filter(|(_, target)| self.cold_blocks[*target]).count();
+            let none_cold = cold_count == 0;
+            let all_cold = cold_count == targets.iter().len();
+            if (none_cold && (!otherwise_cold || otherwise_unreachable))
+                || (all_cold && (otherwise_cold || otherwise_unreachable))
+            {
+                // All targets have the same weight,
+                // or `otherwise` is unreachable and it's the only target with a different weight.
+                bx.switch(
+                    discr_value,
+                    helper.llbb_with_cleanup(self, targets.otherwise()),
+                    target_iter
+                        .map(|(value, target)| (value, helper.llbb_with_cleanup(self, target))),
+                );
+            } else {
+                // Targets have different weights
+                bx.switch_with_weights(
+                    discr_value,
+                    helper.llbb_with_cleanup(self, targets.otherwise()),
+                    otherwise_cold,
+                    target_iter.map(|(value, target)| {
+                        (value, helper.llbb_with_cleanup(self, target), self.cold_blocks[target])
+                    }),
+                );
+            }
         }
     }
 

--- a/compiler/rustc_codegen_ssa/src/mir/mod.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/mod.rs
@@ -502,14 +502,25 @@ fn find_cold_blocks<'tcx>(
     for (bb, bb_data) in traversal::postorder(mir) {
         let terminator = bb_data.terminator();
 
-        // If a BB ends with a call to a cold function, mark it as cold.
-        if let mir::TerminatorKind::Call { ref func, .. } = terminator.kind
-            && let ty::FnDef(def_id, ..) = *func.ty(local_decls, tcx).kind()
-            && let attrs = tcx.codegen_fn_attrs(def_id)
-            && attrs.flags.contains(CodegenFnAttrFlags::COLD)
-        {
-            cold_blocks[bb] = true;
-            continue;
+        match terminator.kind {
+            // If a BB ends with a call to a cold function, mark it as cold.
+            mir::TerminatorKind::Call { ref func, .. }
+            | mir::TerminatorKind::TailCall { ref func, .. }
+                if let ty::FnDef(def_id, ..) = *func.ty(local_decls, tcx).kind()
+                    && let attrs = tcx.codegen_fn_attrs(def_id)
+                    && attrs.flags.contains(CodegenFnAttrFlags::COLD) =>
+            {
+                cold_blocks[bb] = true;
+                continue;
+            }
+
+            // If a BB ends with an `unreachable`, also mark it as cold.
+            mir::TerminatorKind::Unreachable => {
+                cold_blocks[bb] = true;
+                continue;
+            }
+
+            _ => {}
         }
 
         // If all successors of a BB are cold and there's at least one of them, mark this BB as cold

--- a/compiler/rustc_codegen_ssa/src/traits/builder.rs
+++ b/compiler/rustc_codegen_ssa/src/traits/builder.rs
@@ -110,6 +110,20 @@ pub trait BuilderMethods<'a, 'tcx>:
         else_llbb: Self::BasicBlock,
         cases: impl ExactSizeIterator<Item = (u128, Self::BasicBlock)>,
     );
+
+    // This is like `switch()`, but every case has a bool flag indicating whether it's cold.
+    //
+    // Default implementation throws away the cold flags and calls `switch()`.
+    fn switch_with_weights(
+        &mut self,
+        v: Self::Value,
+        else_llbb: Self::BasicBlock,
+        _else_is_cold: bool,
+        cases: impl ExactSizeIterator<Item = (u128, Self::BasicBlock, bool)>,
+    ) {
+        self.switch(v, else_llbb, cases.map(|(val, bb, _)| (val, bb)))
+    }
+
     fn invoke(
         &mut self,
         llty: Self::Type,

--- a/tests/codegen/intrinsics/cold_path2.rs
+++ b/tests/codegen/intrinsics/cold_path2.rs
@@ -1,0 +1,36 @@
+//@ compile-flags: -O
+#![crate_type = "lib"]
+#![feature(core_intrinsics)]
+
+use std::intrinsics::cold_path;
+
+#[inline(never)]
+#[no_mangle]
+pub fn path_a() {
+    println!("path a");
+}
+
+#[inline(never)]
+#[no_mangle]
+pub fn path_b() {
+    println!("path b");
+}
+
+#[no_mangle]
+pub fn test(x: Option<bool>) {
+    if let Some(_) = x {
+        path_a();
+    } else {
+        cold_path();
+        path_b();
+    }
+
+    // CHECK-LABEL: @test(
+    // CHECK: br i1 %1, label %bb2, label %bb1, !prof ![[NUM:[0-9]+]]
+    // CHECK: bb1:
+    // CHECK: path_a
+    // CHECK: bb2:
+    // CHECK: path_b
+}
+
+// CHECK: ![[NUM]] = !{!"branch_weights", {{(!"expected", )?}}i32 1, i32 2000}

--- a/tests/codegen/intrinsics/cold_path3.rs
+++ b/tests/codegen/intrinsics/cold_path3.rs
@@ -1,0 +1,87 @@
+//@ compile-flags: -O
+#![crate_type = "lib"]
+#![feature(core_intrinsics)]
+
+use std::intrinsics::cold_path;
+
+#[inline(never)]
+#[no_mangle]
+pub fn path_a() {
+    println!("path a");
+}
+
+#[inline(never)]
+#[no_mangle]
+pub fn path_b() {
+    println!("path b");
+}
+
+#[inline(never)]
+#[no_mangle]
+pub fn path_c() {
+    println!("path c");
+}
+
+#[inline(never)]
+#[no_mangle]
+pub fn path_d() {
+    println!("path d");
+}
+
+#[no_mangle]
+pub fn test(x: Option<u32>) {
+    match x {
+        Some(0) => path_a(),
+        Some(1) => {
+            cold_path();
+            path_b()
+        }
+        Some(2) => path_c(),
+        Some(3) => {
+            cold_path();
+            path_d()
+        }
+        _ => path_a(),
+    }
+
+    // CHECK-LABEL: @test(
+    // CHECK: switch i32 %1, label %bb1 [
+    // CHECK: i32 0, label %bb6
+    // CHECK: i32 1, label %bb5
+    // CHECK: i32 2, label %bb4
+    // CHECK: i32 3, label %bb3
+    // CHECK: ], !prof ![[NUM1:[0-9]+]]
+}
+
+#[no_mangle]
+pub fn test2(x: Option<u32>) {
+    match x {
+        Some(10) => path_a(),
+        Some(11) => {
+            cold_path();
+            path_b()
+        }
+        Some(12) => {
+            unsafe { core::intrinsics::unreachable() };
+            path_c()
+        }
+        Some(13) => {
+            cold_path();
+            path_d()
+        }
+        _ => {
+            cold_path();
+            path_a()
+        }
+    }
+
+    // CHECK-LABEL: @test2(
+    // CHECK: switch i32 %1, label %bb1 [
+    // CHECK: i32 10, label %bb5
+    // CHECK: i32 11, label %bb4
+    // CHECK: i32 13, label %bb3
+    // CHECK: ], !prof ![[NUM2:[0-9]+]]
+}
+
+// CHECK: ![[NUM1]] = !{!"branch_weights", i32 2000, i32 2000, i32 1, i32 2000, i32 1}
+// CHECK: ![[NUM2]] = !{!"branch_weights", i32 1, i32 2000, i32 1, i32 1}

